default.nix

```
@@ -1,116 +1,150 @@
+    pname = "Agda-ledger";
     version = "0.1";
-    src = "${agda-ledger}";
-    meta = {};
-    buildInputs = [ agdaWithPkgs ];
-    buildPhase = "";
-    installPhase = ''
-      mkdir -p $out
-      agda -c --ghc-dont-call-ghc --compile-dir $out Ledger/Foreign/HSLedger.agda
-      cp Ledger/HSLedgerTest.hs $out
-      cp Ledger/agda-ledger-executable-spec.cabal $out
-      # Append all the modules generated by MAlonzo to the cabal file
-      find $out/MAlonzo -name "*.hs" -print | sed "s#^$out/#        #;s#\.hs##;s#/#.#g" >> $out/agda-ledger-executable-spec.cabal
-    '';
+    src = ./src;
+    meta = { };
+    buildInputs = deps;
+    postInstall = "cp -r latex $out";
+    extraExtensions = [ "hs" "cabal" ];
   };
-in {
-  agda = agdaWithPkgs;
-  agda-ledger = agda-ledger;
-  agda-ledger-executable-spec = haskellPackages.callCabal2nix "agda-ledger-executable-spec" "${ledger-hs-src}" {};
 
-  ledger-docs = stdenv.mkDerivation {
-    pname = "simple-utxo-ledger-docs";
-    version = "0.1";
-    src = "${agda-ledger}";
-    meta = {};
-    buildInputs = [ agdaWithPkgs (texlive.combine {
-      inherit (texlive)
-        scheme-small
-        xits
-        collection-latexextra
-        collection-latexrecommended
-        collection-mathscience
-        bclogo
-        latexmk;
-    })];
-    buildPhase = ''
-      ls Ledger/*.lagda | xargs -n 1 agda --latex
-      cd latex && latexmk -xelatex Ledger/Ledger.tex && cd ..
-    '';
-    installPhase = ''
-      mkdir -p $out
-      agda --html --html-dir $out/html Ledger/Ledger.lagda
-      cp latex/Ledger.pdf $out
-    '';
+  hsSrc = { dir, agdaLedgerPath, hsMainPath }:
+    stdenv.mkDerivation {
+      pname = "Agda-ledger-${dir}-hs-src";
+      version = "0.1";
+      src = "${agdaLedger}";
+      meta = { };
+      buildInputs = [ agdaWithPkgs ];
+      buildPhase = "";
+      installPhase = ''
+        mkdir -p $out
+        agda -c --ghc-dont-call-ghc --compile-dir $out ${agdaLedgerPath}
+        cp ${hsMainPath} $out
+        cp ${dir}/agda-ledger-executable-spec.cabal $out
+        # Append all the modules generated by MAlonzo to the cabal file
+        find $out/MAlonzo -name "*.hs" -print | sed "s#^$out/#        #;s#\.hs##;s#/#.#g" >> $out/agda-ledger-executable-spec.cabal
+      '';
+    };
+
+  ledgerHsSrc = hsSrc {
+    dir = "Ledger";
+    agdaLedgerPath = "Ledger/Foreign/HSLedger.agda";
+    hsMainPath = "Ledger/HSLedgerTest.hs";
+  };
+
+  ledgerExecutableSpec = haskellPackages.callCabal2nix "Agda-ledger-executable-spec" "${ledgerHsSrc}" { };
+
+  ledgerHsSrcMidnight = hsSrc {
+    dir = "MidnightExample";
+    agdaLedgerPath = "MidnightExample/HSLedger.agda";
+    hsMainPath = "MidnightExample/Main.hs";
+  };
+
+  ledgerExecutableSpecMidnight = haskellPackages.callCabal2nix "Agda-ledger-executable-spec-midnight" "${ledgerHsSrcMidnight}" { };
+
+  docs = { dir }:
+    stdenv.mkDerivation {
+      pname = "Simple-utxo-ledger-${dir}-docs";
```

That name is pretty old, I think this would be better
```suggestion
      pname = "${dir}-docs";
```